File: src/com/github/stromland/piano/services/Piano.kt
package com.github.stromland.piano.services

import com.github.stromland.piano.models.PianoKey
import kotlin.math.ceil

object Piano {
    val pianoKeys = listOf(
        PianoKey("C"),
        PianoKey("C#", true),
        PianoKey("D"),
        PianoKey("D#", true),
        PianoKey("E"),
        PianoKey("F"),
        PianoKey("F#", true),
        PianoKey("G"),
        PianoKey("G#", true),
        PianoKey("A"),
        PianoKey("Bb", true),
        PianoKey("B")
    )

    fun getPianoKeyIndex(note: String): Int {
        return pianoKeys
            .mapIndexed { index, pianoKey -> pianoKey.note to index }
            .find { it.first == note }
            ?.second ?: throw IllegalArgumentException("Not a valid note ($note)")
    }

    fun getPianoKeySets(sets: Int) = List(sets) { pianoKeys }.reduce { allKeys, keys -> allKeys + keys }

    fun getPianoKeys(keys: Int): List<PianoKey> {
        val sets = ceil(keys / pianoKeys.size.toDouble()).toInt()
        return getPianoKeySets(sets).slice(0 until keys)
    }

    fun pressKeys(pianoKeys: List<PianoKey>, keysToPress: List<Int>): List<PianoKey> {
        return pianoKeys.mapIndexed { index, pianoKey ->
            keysToPress.find { index == it }?.let { pianoKey.copy(pressed = true) } ?: pianoKey
        }
    }
}

File: src/services/Piano.ts
import { PianoKey } from "../models/PianoKey";

export const PIANO_NOTES = [
  "C",
  "C#",
  "D",
  "D#",
  "E",
  "F",
  "F#",
  "G",
  "G#",
  "A",
  "Bb",
  "B"
] as const;

export type Note = typeof PIANO_NOTES[number];

export class Piano {
  public static keys: PianoKey[] = PIANO_NOTES.map(note => ({
    note,
    black: note.length === 2,
    pressed: false
  }));

  public static getPianoKeyIndex(note: Note): number {
    return this.keys.findIndex(it => it.note === note);
  }

  public static getNoteFromIndex(index: number): Note {
    return PIANO_NOTES[index % 12];
  }

  public static getPianoKeySets(sets: number): PianoKey[] {
    return Array(sets)
      .fill(this.keys)
      .reduce((allKeys, keys) => [...allKeys, ...keys]);
  }

  public static getPianoKeys(keys: number): PianoKey[] {
    const sets = Math.ceil(keys / this.keys.length);
    return this.getPianoKeySets(sets).slice(0, keys);
  }

  public static pressKeys(keys: PianoKey[], keysToPress: number[]): PianoKey[] {
    return keys.map((key, index) =>
      keysToPress.includes(index) ? { ...key, pressed: true } : key
    );
  }
}
